Add AWS Serverless Forwarder to Obs Guide

ollyhowell opened the pull request:
Preview: https://observability-docs_2214.docs-preview.app.elstc.co/guide/en/observability/master/aws-elastic-serverless-forwarder.html

Open issues and questions:

- [x] Do we have the positioning within left-hand nav / table of contents, and _cut_ of sub-pages ok? Going from a single README to parent and child pages, is this best way to present content? e.g. see JSON discovery topic, or configuration options in addition to Configure... -- See comments but will move to Send data... section for now
- [x] Need update for monitor-logs.html parent page (or wherever AWS content is placed) -- redundant for now
- [x] Add extra images from https://www.elastic.co/blog/elastic-and-aws-serverless-application-repository-speed-time-to-actionable-insights-with-frictionless-log-ingestion-from-amazon-s3 e.g. https://static-www.elastic.co/v3/assets/bltefdd0b53724fa2ce/blt10380108dc86421b/61f82963c0b94b3fdbdd1fc3/Image_3_Step4_InstallAWSAssets.png and https://static-www.elastic.co/v3/assets/bltefdd0b53724fa2ce/blt687784ea6bddd0af/61f842135865427ee15b8833/Image_5-1Sarconfig_(1).png -- SAR install option useful, other will date quickly
- [x] Run linter and apply recommendations (check o/s attributes in particular)
- [ ] ~Some content 'dumped' in Notes should be better integrated / restructured~
- [ ] ~Create tables with some of the field/setting lists?~
- [x] Add keywords and description to new topics?
- [x] Cut back content in https://github.com/elastic/elastic-serverless-forwarder/blob/main/docs/README-AWS.md? Work out overlap with what needs to be integrated within AWS docs (currently whole README!)
- [x] Peer review from obs writer...partly done, will cover during cloud obs overhaul in new year
- [x] Work out which versions to backport to

Closes https://github.com/elastic/observability-docs/issues/1918

Note: I have not (yet) tried to get this working from scratch or stepped thru the detailed instructions...
UPDATE: now working! Making some edits based on experience...

apmmachine commented:
A documentation preview will be available soon:</br><ul><li>📚 [HTML diff](https://observability-docs_2214.docs-preview.app.elstc.co/diff)</li><li>📘 [Fleet and Elastic Agent Guide](https://observability-docs_2214.docs-preview.app.elstc.co/guide/en/fleet/master/index.html)</li><li>📙 [Observability Guide](https://observability-docs_2214.docs-preview.app.elstc.co/guide/en/observability/master/index.html)</li></ui>
<!--COMMENT_GENERATED_WITH_ID_comment.id-->

ravikesarwani commented:
> Do we have the positioning within left-hand nav / table of contents, and cut of sub-pages ok? Going from a single README to parent and child pages, is this best way to present content? e.g. see JSON discovery topic, or configuration options in addition to Configure...

Thanks @ollyhowell  for your time and effort on this. This looks really good and much better organized. 
Some comments to think about.

**Placement of the topic under “Log monitoring”:**
I was unsure if this is the right positioning as all the current topics in “Log monitoring” are focused on using the logs data rather than how to ingest log data. Maybe “AWS Elastic Serverless Forwarder” makes more sense to be put under “Send data to Elasticsearch”.

**Sub pages:** 
Maybe we can think about simplifying the sub topics as “Deploy=>Configure=>Transform=>Troubleshooting”.
This is how I think customers most likely will think about the flow as well.

Maybe we can shorten it to just “Deploy” instead of “Deploy Elastic Serverless Forwarder”.
Maybe we can shorten it to just “Configure” instead of “Configure Elastic Serverless Forwarder”.
Merge “Route AWS service logs” sub topic into “Configure” itself rather than as a separate sub-topic.
Add a new topic called “Transform” that will include sub-topics like “Use custom tags, Define include/exclude, JSON content discovery, manage multiline messages”.
Maybe change “Troubleshooting and error handling” to just “Troubleshooting” to keep it simple.

ravikesarwani commented:
cc: @tommyers-elastic

ollyhowell commented:
> **Placement of the topic under “Log monitoring”:** I was unsure if this is the right positioning as all the current topics in “Log monitoring” are focused on using the logs data rather than how to ingest log data. Maybe “AWS Elastic Serverless Forwarder” makes more sense to be put under “Send data to Elasticsearch”.

We could move it under Send data... topic quite easily. We know that section will change and we will build out an AWS section so it's really just picking the right location for it now.

> **Sub pages:** Maybe we can think about simplifying the sub topics as “Deploy=>Configure=>Transform=>Troubleshooting”. This is how I think customers most likely will think about the flow as well.
> 
> Maybe we can shorten it to just “Deploy” instead of “Deploy Elastic Serverless Forwarder”. Maybe we can shorten it to just “Configure” instead of “Configure Elastic Serverless Forwarder”. Merge “Route AWS service logs” sub topic into “Configure” itself rather than as a separate sub-topic. Add a new topic called “Transform” that will include sub-topics like “Use custom tags, Define include/exclude, JSON content discovery, manage multiline messages”. Maybe change “Troubleshooting and error handling” to just “Troubleshooting” to keep it simple.

I was thinking similar about shortening titles - or at very least making the nav/TOC entry the abbreviated version as the longer 'Deploy Elastic Serverless Forwarder' title and matching deploy-elastic-serverless-forwarder filename will produce good search results for any searches e.g. deploy elastic serverless forwarder.

Tip about moving Route into Configure helps, and I can combine the others into a Transform topic as suggested.

ollyhowell commented:
Not moved the section yet but with topics combined and relabelled:

![image](https://user-images.githubusercontent.com/89071729/193843612-87a405fb-8174-4a99-8255-831eb704d931.png)

ravikesarwani commented:
@aspacca Can you please make sure to review the doc for technical correctness and provide your okay. Thanks!

aspacca commented on docs/en/observability/aws-serverless-manage-multiline-messages.asciidoc:
```diff
@@ -0,0 +1,241 @@
+
+This configuration merges any line that ends with the `\` character with the line that follows it.
+
+NOTE: Note that you should escape the opening square bracket (`[`) in the regular expression, because it specifies a character class i.e. a set of characters that you wish to match. You also have to escape the backslash (`\`) used for escaping the opening square bracket as raw strings are not used. Thus, `^\\[` will produce the required regular expression upon compiling.
```
the `NOTE` comes from the different section
the original one is
```
*Notes*: beware that we should escape the backslash we want to match a string literals (`\`) twice, since we are not using raw string. The final result `\\\\$` will produce the regular expression to be compiled properly.
```

ollyhowell replied:
Good spot, thanks - was trying to make these notes consistent. I've fixed in new transform- topic that now contains the multiline content.

aspacca commented on docs/en/observability/aws-serverless-use-tags-filters.asciidoc:
```diff
@@ -0,0 +1,89 @@
+
+`["forwarded", "generic", "tag1", "tag2", "tag3"]`
+
+// where do first two tags come from?
```
`forwarded` tag is always appended 
`generic` tag comes from the dataset

aspacca commented on docs/en/observability/aws-serverless-use-tags-filters.asciidoc:
```diff
@@ -0,0 +1,89 @@
+====
+Both config parameters are optional, and can be set independently of each other. In terms of rule precedence, the exclude filter is applied first and then the include filter.
+
+//check "so exclude takes precedence if both are supplied"?
```
yes

aspacca commented on docs/en/observability/aws-serverless-use-tags-filters.asciidoc:
```diff
@@ -0,0 +1,89 @@
+
+Messages are scanned for terms that match the defined filters. Anchoring at the beginning of the string must be done explicitly with the `^` (caret) special character.
+
+// what is anchoring? Expand?
```
forcing the matching to start from the beginning of the message: anchoring at the beginning/anchoring at the end (`^`/`$`).

"line anchors" is a common definition in regex, not sure if we need to explain :)

ollyhowell replied:
Thanks, have revised slightly

aspacca commented on docs/en/observability/configure-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,255 @@
+----
+
+In addition to these basic permissions, the following permissions are added when the function is created:
+// <!-- how? by whom?-->:
```
automatically: by the cloudformation template of the forwarder

the user provides ARNs of AWS resources the Forwarder will "interact" with and the CF template generate the proper IAM role with the proper IAM policies

ollyhowell replied:
Thanks, have added

ollyhowell commented:
@elasticmachine run elasticsearch-ci/docs rebuild

ravikesarwani commented:
@ollyhowell Is there a way I can see the latest preview?

ollyhowell commented:
Yeah, was just waiting on the CI/CD kicking in...had to bump it with a comment but it seems to be rebuilding...will update with url soon

ollyhowell commented:
Ok, preview back! @ravikesarwani 

Same as listed at top of issue: https://observability-docs_2214.docs-preview.app.elstc.co/guide/en/observability/master/aws-elastic-serverless-forwarder.html

ravikesarwani commented on docs/en/observability/aws-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,99 @@
+= Overview
+The Elastic Serverless Forwarder can forward {aws} data to self-managed or cloud-hosted Elastic environments. It supports the following inputs:
+
+- Amazon S3 SQS (Simple Queue Service) event notifications
```
Amazon S3 (via SQS event notifications)

ravikesarwani commented on docs/en/observability/aws-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,99 @@
+- Amazon S3 SQS (Simple Queue Service) event notifications
+- Amazon Kinesis Data Streams
+- Amazon CloudWatch Logs subscription filters
+- Direct Amazon SQS message payload
```
Amazon SQS message payload

ravikesarwani commented on docs/en/observability/aws-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,99 @@
+[role="screenshot"]
+image::aws-serverless-lambda-flow.png[AWS Lambda flow]
+
+The Lambda deployment sets up an SQS queue automatically. This is used to trigger a new function invocation that Lambda uses to continue from exactly where the last function run was terminated. By default a Lambda function runs for a maximum of 15 minutes. When processing event data there’s a possibility that the function may be exited by {aws} in the middle of processing. The function handles th ...
```
The Lambda deployment sets up a continuing SQS queue automatically.

ollyhowell replied:
Think I edited that original reference out lol - what do you mean by "continuing"?
I don't think Amazon refer to a "continuing SQS queue" in their docs so trying to understand what this means in this context e.g. "persistent"?

ravikesarwani replied:
Yes, its an SQS queue but we call it continuing queue (our name). This is what we are showing in our diagram as well.
This helps us to differentiate and refer to the specific queue (as you can see there are many) by its purpose (helps with checkpointing and continue where where we left of in the last run).

ollyhowell replied:
Ah, thanks - I see it in the diagram now. Updating locally, will push another commit and then ask for a peer review from an Obs writer.

ravikesarwani commented on docs/en/observability/aws-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,99 @@
+The Lambda function also supports writing directly to an index, alias, or custom data stream. This enables existing {es} users to re-use index templates, ingest pipelines, or dashboards that are already created and connected to business processes.
+
+The best way to get started is to install the appropriate https://docs.elastic.co/en/integrations[integrations] via {kib}. These integrations include pre-built dashboards, ingest node configurations, and other assets that help you get the most out of the data you ingest. The integrations use https://www.elastic.co/guide/en/elasticsearch/reference/current/data-streams.html[data streams] with speci ...
+
```
Add assets after integration.
The best way to get started is to install the appropriate https://docs.elastic.co/en/integrations[integrations] **assets** via {kib}. These integrations **assets** include pre-built dashboards, ingest node configurations, and other assets that help you get the most out of the data you ingest. The integrations use https://www.elastic.co/guide/en/elasticsearch/reference/current/data-streams.html[data streams] with specific https://www.elastic.co/blog/an-introduction-to-the-elastic-data-stream-naming-scheme[naming conventions] that provide you with more granular controls and flexibility on managing data ingestion.

ravikesarwani commented on docs/en/observability/aws-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,99 @@
+[discrete]
+[[aws-serverless-forwarder-inputs-direct]]
+== Direct Amazon SQS message payload
+
```
We can get rid of "Direct" and keep it simple as "Amazon SQS message payload".
Can we also change the order to match the top and the same exact names. Order: Amazon S3=> Kinesis => cloudwatch log =>SQS.

ravikesarwani commented on docs/en/observability/aws-serverless-transform.asciidoc:
```diff
@@ -0,0 +1,412 @@
+
+[[aws-serverless-transform]]
+= Transform
+
```
Can we put the various transform topics as links in the top as well (like table of content)? Right now the user has to scan the entire page to know what kind of data transformation features are available on the page and its a long page so information is not easy to find.

ollyhowell replied:
e.g.
You can transform or enrich data from {aws} as it is parsed by the Elastic Serverless Forwarder. This includes examples such as using tags and filters to organise your data and exclude specific messages, automatically discovering and collecting JSON content, and managing multiline messages effectively.

colleenmcginnis commented on docs/en/observability/aws-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,99 @@
+
+Any exception or fail scenarios related to ingestion are handled by Lambda gracefully using a _Replay queue_. Data in the replay queue is stored as individual events. Lambda keeps track of any failed events and writes them to a replay queue that can then be consumed by adding an additional SQS trigger via Lambda.
+
+You can use the <<sample-s3-config-file,config.yaml>> file to configure the service for each input type, including information such as SQS queue ARN and {es} connection details. You can create multiple input sections within the configuration file to map different inputs to specific log types.
```
>SQS queue ARN

I don't know what this means. Do we want to assume the reader will know or should we explain?

dedemorton replied:
AWS users will know what this means.

colleenmcginnis commented on docs/en/observability/aws-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,99 @@
+
+You can use the <<sample-s3-config-file,config.yaml>> file to configure the service for each input type, including information such as SQS queue ARN and {es} connection details. You can create multiple input sections within the configuration file to map different inputs to specific log types.
+
+The Lambda function also supports writing directly to an index, alias, or custom data stream. This enables existing {es} users to re-use index templates, ingest pipelines, or dashboards that are already created and connected to business processes.
```
I don't really know what "are connected to business processes" means. Does that just mean they are already in use? 

```suggestion
The Lambda function also supports writing directly to an index, alias, or custom data stream. This enables existing {es} users to re-use index templates, ingest pipelines, or dashboards that already exist and are connected to business processes.
```

ollyhowell replied:
Yes but I agree some examples or details would help here... @ravikesarwani could you help with this one?

colleenmcginnis commented on docs/en/observability/aws-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,99 @@
+
+The Lambda function also supports writing directly to an index, alias, or custom data stream. This enables existing {es} users to re-use index templates, ingest pipelines, or dashboards that are already created and connected to business processes.
+
+The best way to get started is to install the appropriate integration assets via {kib}. These integration assets include pre-built dashboards, ingest node configurations, and other assets that help you get the most out of the data you ingest. The integrations use https://www.elastic.co/guide/en/elasticsearch/reference/current/data-streams.html[data streams] with specific https://www.elastic.co/bl ...
```
Should this be with the `= Get started` section below?

ollyhowell replied:
Yeah, it is a little out of place / overlapping with either the Get started _section_ or the Deploy topic itself - I left it in as it was a useful CTA inside quite a dense introductory section but I'll try moving it into Deploy I think...

colleenmcginnis commented on docs/en/observability/aws-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,99 @@
+[[aws-serverless-forwarder-inputs-kinesis]]
+== Amazon Kinesis Data Streams
+
+The Lambda function can ingest logs contained in the payload of a Kinesis Data Stream record and send them to Elastic. The Kinesis Data Stream serves as a trigger for the Lambda function. When a new record gets written to a Kinesis Data Stream, the Lambda function triggers.
```
```suggestion
The Lambda function can ingest logs contained in the payload of a Kinesis Data Stream record and send them to Elastic. The Kinesis Data Stream serves as a trigger for the Lambda function. When a new record gets written to a Kinesis Data Stream, it triggers the Lambda function.
```

colleenmcginnis commented on docs/en/observability/aws-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,99 @@
+[[aws-serverless-forwarder-inputs-cloudwatch]]
+== Amazon CloudWatch Logs subscription filters
+
+The Lambda function can ingest logs contained in the message payload of CloudWatch Logs events and send them to Elastic. The CloudWatch Logs service serves as a trigger for the Lambda function. When a new record gets written to a Kinesis data stream, the Lambda function triggers.
```
```suggestion
The Lambda function can ingest logs contained in the message payload of CloudWatch Logs events and send them to Elastic. The CloudWatch Logs service serves as a trigger for the Lambda function. When a new record gets written to a Kinesis data stream, it triggers the Lambda function.
```

aspacca replied:
beware we are mixing cloudwatch logs and kinesis data stream here in the description:
```suggestion
The Lambda function can ingest logs contained in the message payload of CloudWatch Logs events and send them to Elastic. The CloudWatch Logs service serves as a trigger for the Lambda function. When a new event gets written to a CloudWatch Logs log stream, it triggers the Lambda function.
```

about:
> it triggers the Lambda function

indeed it's a little different than this. multiple events written to the log stream can trigger the lambda function "in batch".

colleenmcginnis commented on docs/en/observability/aws-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,99 @@
+[[aws-serverless-forwarder-inputs-direct]]
+== Amazon SQS message payload
+
+The Lambda function can ingest logs contained within the payload of an Amazon SQS body record and send them to Elastic. The SQS queue serves as a trigger for the Lambda function. When a new record gets written to an SQS queue, the Lambda function triggers.
```
```suggestion
The Lambda function can ingest logs contained within the payload of an Amazon SQS body record and send them to Elastic. The SQS queue serves as a trigger for the Lambda function. When a new record gets written to an SQS queue, it triggers the Lambda function.
```

colleenmcginnis reviewed (commented):
I struggled a bit with the passive voice and my limited knowledge of AWS vocabulary. That combination made it difficult for me to understand how the pieces fit together, but it may be that we're comfortable assuming the reader has more familiarity with AWS than I do. 

I'm going to pause here and make sure this _kind_ of feedback is helpful before moving onto the instructional content.

dedemorton commented on docs/en/observability/aws-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,99 @@
+:aws: AWS
+
+[[aws-elastic-serverless-forwarder]]
+= Elastic Serverless Forwarder
+
+// should we expand titleabbrev to AWS or Amazon Web Services Elastic Serverless Forwarder? What is correct nomenclature?
+
+++++
+<titleabbrev>AWS Elastic Serverless Forwarder</titleabbrev>
+++++
```
Colleen's suggestion is inline with their trademark guidelines.

dedemorton replied:
Nothing here that titleabbrev is also used for generated link text.

ollyhowell replied:
> Or if the name of the product is "Elastic Serverless Forwarder" we could use something like "Elastic Serverless Forwarder for Amazon Web Services (AWS)".

We definitely need clarity on naming and how we connect it to AWS. I like this suggestion as it handles possible developments like Elastic Serverless Forwarder for BlahBlah (I know that other cloud platforms implement differently but it's not hard to imagine same concept being used for other cloud providers).

As we don't have the **Cloud observability** section yet, and the AWS bit is needed to make the context clear, I'd experimented with using titleabbrev to see how the menu item looks. I could see us having a longer page title and filenames but using the abbrev to shorten (e.g. AWS rather than Amazon Web Services). What do you think to the naming point @ravikesarwani?

ollyhowell replied:
Any thoughts on the naming issue @ravikesarwani ?

ravikesarwani replied:
Using AWS in the titleabbrev is fine, so something like "Elastic Serverless Forwarder for AWS" should be okay from my perspective (trying to keep it smaller)

dedemorton reviewed (commented):
Just a few responses to Colleen's edits. I still plan to test and offer suggestions.

colleenmcginnis commented on docs/en/observability/configure-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,278 @@
+
+In addition to these basic permissions, the following permissions are added when the function is created by the Cloudformation template of the function:
+
+* For SQS queue resources specified in the `SQS_CONTINUE_URL` and `SQS_REPLAY_URL` environment variables, the following action is allowed:
+
+  `sqs:SendMessage`
```
Is there a specific reason to format these list items like this? Could we just use:

```suggestion
* For SQS queue resources specified in the `SQS_CONTINUE_URL` and `SQS_REPLAY_URL` environment variables, the following action is allowed: `sqs:SendMessage`.
```

ollyhowell replied:
I think the last one included two allowed actions on separate lines but the rest are just single items and I agree it's better without the extra line break

colleenmcginnis commented on docs/en/observability/configure-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,278 @@
+
+  `s3:GetObject`
+
+* For every S3 bucket resource that SQS queues are receiving notifications from, the following action is allowed on the S3 buckets:
+
+  `s3:ListBucket`
+
+* For every S3 bucket resource that SQS queues are receiving notifications from, the following action is allowed on the S3 buckets' keys:
+
+  `s3:GetObject`
```
Is this still accurate? 

```suggestion
* For every S3 bucket resource sending notifications to SQS queues, the following action is allowed on the S3 buckets:

  `s3:ListBucket`

* For every S3 bucket resource sending notifications to SQS queues, the following action is allowed on the S3 buckets' keys:

  `s3:GetObject`
```

ollyhowell replied:
I think so (made changes locally) though tagging @aspacca for sanity check

aspacca replied:
oki for me

colleenmcginnis commented on docs/en/observability/configure-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,278 @@
+
+  `s3:GetObject`
+
+* For every Secret Manager secret that you want to refer in the yaml configuration file, the following action is allowed
```
```suggestion
* For every Secret Manager secret that you want to refer in the yaml configuration file, the following action is allowed:
```

I'd also choose one style and stick with it across files:

* `config.yaml` file
* config YAML file
* YAML configuration file

colleenmcginnis commented on docs/en/observability/configure-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,278 @@
+
+  `secretsmanager:GetSecretValue`
+
+* For every decrypt key that's not the default one used to encrypt your Secret Manager secrets with, the following action is allowed:
```
The first part of this sentence is quite difficult to parse.

ollyhowell replied:
Yeah, agreed - how about?

> * Excepting the default key used to encrypt your Secret Manager secrets with, the following action is allowed for every decrypt key: `kms:Decrypt`

colleenmcginnis commented on docs/en/observability/configure-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,278 @@
+
+  `kms:Decrypt`
+
+* If any CloudWatch Logs log groups are set as Lambda inputs, the following actions are allowed for the resource
+
+  `arn:aws:logs:%AWS_REGION%:%AWS_ACCOUNT_ID%:log-group:*:*`:
+  `logs:DescribeLogGroups`
```
```suggestion
* If any CloudWatch Logs log groups are set as Lambda inputs, the following actions are allowed for the resource:
** `arn:aws:logs:%AWS_REGION%:%AWS_ACCOUNT_ID%:log-group:*:*`
** `logs:DescribeLogGroups`
```

dedemorton commented on docs/en/observability/aws-serverless-transform.asciidoc:
```diff
@@ -0,0 +1,416 @@
+:aws: AWS
+
+[[aws-serverless-transform]]
+= Transform AWS data
```
These are all things you need to do in that config.yaml before you upload it to an s3 bucket. It seems to me that all the config-related info should be in the sample place rather than buried under a category like "transform".

ollyhowell replied:
Maybe the Configure and Transform split isn't the best...have added a short note to the Deploy topic for now while we work out how to integrate/restructure the additional configure and transform details...

dedemorton commented on docs/en/observability/configure-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,278 @@
+
+When you provide the ARNs of {aws} resources the forwarder will interact with, the Cloudformation template will create the correct IAM role with the appropriate IAM policies.
+
+You can view the execution role associated with your function from the **Configuration -> Permissions** section. By default this role starts with the name **serverlessrepo-elastic-se-ElasticServerlessForward-**. When the role is created, a custom policy is added to grant Lambda minimum permissions to be able to use the configured SQS queue, S3 buckets, Kinesis data stream, CloudWatch Logs log gro ...
```
Could not test this step... serverlessrepo-elastic-se-ElasticServerlessForward does not exist for me under Lambda > Functions. I see this, but I'm using a shared environment.

![image](https://user-images.githubusercontent.com/14206422/195956541-1c23b5e5-617e-4d7e-a2ee-8129c88b99f8.png)

aspacca replied:
you are right @dedemorton 

the proper name is now:
`**serverlessrepo-elastic-se-ApplicationElasticServer-**`

I forgot to update it in the readme on the repo when it was renamed

dedemorton commented on docs/en/observability/deploy-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,164 @@
+      |`ELASTIC_APM_SECRET_TOKEN` | token  |
+      |`ELASTIC_APM_SERVER_URL`	  | url    |
+
+[[aws-serverless-forwarder-deploy-cloudformation]]
```
What happens next? Will users see the data streams in Elasticsearch? I tried this section, which should have been the easiest for me, but must have messed something up because I'm not getting any logs in Elasticsearch.

aspacca replied:
@dedemorton 

>Could not test this step... serverlessrepo-elastic-se-ElasticServerlessForward does not exist for me under Lambda > Functions. I see this, but I'm using a shared environment.

how did you deployed the forwarder? according to the method (console, terraform, cloudformation) the prefix before "MacroElastic..." and "ApplicationElastic..." may change.
You could try to refresh the list of lambdas and order by last updated.



> What happens next? Will users see the data streams in Elasticsearch? I tried this section, which should have been the easiest for me, but must have messed something up because I'm not getting any logs in Elasticsearch.

it should. I wonder if the deployment of the forwarder was successful at all in your test.
if you go to the Cloudformation console can you see the stack? what's its status?
you should have something like this:
<img width="1380" alt="Schermata 2022-10-15 alle 09 06 08" src="https://user-images.githubusercontent.com/221530/195959630-29eaaf2b-6ad7-481e-8180-a38901873723.png">

being a shared environment, if a forwarder was already deployed, including the macro, you cannot deploy another one in the same region

dedemorton replied:
@aspacca I used the console. I found the instructions about the configuration and application settings to be pretty confusing, so I'm sure I must have configured things incorrectly.

ollyhowell replied:
I had another go at getting this working today, with mixed results...I can see one function working but not see the data in Elastic yet. I also deleted a function that failed to deploy - perhaps because it is a shared environment and a forwarder was already deployed? i.e. 

> being a shared environment, if a forwarder was already deployed, including the macro, you cannot deploy another one in the same region

I think a short zoom might be useful to ask you some questions and potentially troubleshoot some deployment issues - will ping you and see if we can find a time!

ollyhowell replied:
I got it all working yesterday after a zoom session with Andrea - thank you!

On this point:
> Could not test this step... serverlessrepo-elastic-se-ElasticServerlessForward does not exist for me under Lambda > Functions. I see this, but I'm using a shared environment.

I can see mine but it is named _serverlessrepo-odh-elasti-ApplicationElasticServer-X59MCXID555K_ obviously taking odh from one of my named objects. You could check and see if you see anything named serverlessrep- @dedemorton? Or when you have time we could take a look at your setup together if you're still keen to get it working (perhaps when I push a better version of the docs?)

Have updated that doc line to:

> By default this role starts with the name **serverlessrepo-**.

dedemorton replied:
>You could check and see if you see anything named serverlessrep-

Yes, I do see some functions that begin with serverlessrepo (see the screen captures that I attached above). @dedemorton?

> we could take a look at your setup together if you're still keen to get it working (perhaps when I push a better version of the docs?)

My plate is pretty full for the next couple of weeks. Maybe when I get back from PTO in January?

ollyhowell replied:
Yeah, that would be fine - it's difficult to dip into _all-the-aws-stuff_ with limited time and I think we will have enough review improvements to get this PR merged in before then and review again when we look at AWS and Cloud more completely.

dedemorton reviewed (commented):
I tried to test some of the tasks here, but TBH it's too hard right now. I didn't want to waste a lot of time thrashing on my own. I'd recommend that you get a developer to walk through the steps because I don't think the content is arranged in a very user friendly way right now. IMO you can't get it right without running through the steps to see the workflow.

ollyhowell commented on docs/en/observability/aws-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,99 @@
+= Overview
+The Elastic Serverless Forwarder can forward {aws} data to self-managed or cloud-hosted Elastic environments. It supports the following inputs:
+
+- Amazon S3 (via SQS (Simple Queue Service) event notifications)
```
Think I was trying to do too much with this bullet - going to simplify the bullet text and explain both terms in the line 47 section. The terms are familiar enough with Amazon users that we can give this first reference a pass.

ollyhowell commented on docs/en/observability/aws-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,99 @@
+[role="screenshot"]
+image::aws-serverless-lambda-flow.png[AWS Lambda flow]
+
+The Lambda deployment sets up an SQS _Continuing queue_ automatically. This is used to trigger a new function invocation that Lambda uses to continue from exactly where the last function run was terminated. By default a Lambda function runs for a maximum of 15 minutes. When processing event data there’s a possibility that the function may be exited by {aws} in the middle of processing. The functi ...
```
Thanks, I like your new suggestion

ollyhowell commented on docs/en/observability/deploy-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,164 @@
+:keywords: serverless
+:description: Deploy the Elastic Serverless Forwarder using Kibana and Amazon's Serverless Application Repository (SAR).
+
+To deploy Elastic Serverless Forwarder, you must perform the following two steps:
+
+. Add {aws} integration in {kib}.
+. Deploy function from {aws} SAR (Serverless Application Repository).
```
Folded into local changes...

ollyhowell commented on docs/en/observability/deploy-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,164 @@
+[[aws-serverless-forwarder-deploy-sar]]
+== Deploy elastic-serverless-forwarder from SAR
+
+There are several deployment methods available via {aws} SAR:
+
+* <<aws-serverless-forwarder-deploy-console>>
+* <<aws-serverless-forwarder-deploy-cloudformation>>
+* <<aws-serverless-forwarder-deploy-terraform>>
```
Good point but I don't know the context here - @ravikesarwani do we have any guidance to add re: deployment methods that will help users decide?

ollyhowell commented on docs/en/observability/deploy-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,164 @@
++
+. Complete the **Application settings** as follows:
+    * `ElasticServerlessForwarderS3ConfigFile` set the value of the S3 URL in the format `s3://bucket-name/config-file-name` to the configuration file for your deployment of Elastic Serverless Forwarder - this will populate the `S3_CONFIG_FILE` environment variable of the deployed Lambda.
+    * `ElasticServerlessForwarderSSMSecrets` add a comma delimited list of {aws} SSM Secrets ARNs referenced in the config yaml file (if any).
+    * `ElasticServerlessForwarderKMSKeys` add a comma delimited list of {aws} KMS Keys ARNs to be used for decrypting {aws} SSM Secrets referenced in the config yaml file (if any).
+    * `ElasticServerlessForwarderSQSEvents` add a comma delimited list of Direct SQS queues ARNs to set as event triggers for the Lambda (if any).
+    * `ElasticServerlessForwarderS3SQSEvents` add a comma delimited list of S3 SQS Event Notifications ARNs to set as event triggers for the Lambda (if any).
+    * `ElasticServerlessForwarderKinesisEvents` add a comma delimited list of Kinesis Data Stream ARNs to set as event triggers for the Lambda (if any).
+    * `ElasticServerlessForwarderCloudWatchLogsEvents` add a comma delimited list of Cloudwatch Logs Log Groups ARNs to set subscription filters on the Lambda (if any).
+    * `ElasticServerlessForwarderS3Buckets` add a comma delimited list of S3 buckets ARNs that are sources for the S3 SQS Event Notifications (if any).
```
> > config yaml
> 
> I recommend consistently using either "`config.yaml` file" or "config YAML file".

Yeah, good call - sticking with `config.yaml` file

> 
> > Cloudwatch Logs Log Groups ARNs
> 
> This might be right, but it looks strange. 🙃

It did and it does - had another look at AWS docs and I think 'Cloudwatch Logs log group ARNs' is used and slightly better...

ollyhowell commented on docs/en/observability/deploy-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,164 @@
+. Deploy function from {aws} SAR (Serverless Application Repository).
+
+[[aws-serverless-forwarder-deploy-kibana]]
+== Add {aws} integration in {kib}
```
Yeah, agree - thanks

ollyhowell commented on docs/en/observability/deploy-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,164 @@
+
+. Go to **Integrations** in {kib} and search for {aws} (or select the **{aws}** category to filter the list).
+. Click the {aws} integration, select **Settings** and click **Install {aws} assets** to install all the {aws} integration assets.
+
```
Good spot - odd the text is different - is it worth amending to **Install AWS assets** @aspacca ?

Small change to mention additional confirmation - wdyt?
> . Click the {aws} integration, select **Settings** and click **Install {aws} assets** and confirm to install all the {aws} integration assets.

aspacca replied:
>is it worth amending to Install AWS assets @aspacca ?

I trust @dedemorton opinion on this :)

ollyhowell replied:
Can we get the button caption updated then?

ollyhowell replied:
Is the button caption going to change to 'Install AWS assets' @aspacca?

aspacca replied:
> Is the button caption going to change to 'Install AWS assets' @aspacca?

we should ping someone in Kibana for this I guess

ollyhowell commented on docs/en/observability/deploy-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,164 @@
+
+Adding integrations from {kib} provides appropriate pre-built dashboards, ingest node configurations, and other assets that help you get the most out of the data you ingest. For more information, refer to the https://docs.elastic.co/en/integrations[Integrations documentation].
+
+// add screenshot?
```
Yeah, that's what I was thinking...will find a screenshot as I think we've used one in other material already.

ollyhowell commented on docs/en/observability/deploy-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,164 @@
+// add screenshot?
+
+[[aws-serverless-forwarder-deploy-sar]]
+== Deploy elastic-serverless-forwarder from SAR
```
They do offer the abbreviation but they seem to normally spell it out so will follow this and your guidance!

ollyhowell commented on docs/en/observability/deploy-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,164 @@
+image::images/aws-serverless-forwarder-create-function.png[Create Elastic Serverless Forwarder Lambda function within SAR]
++
+. Complete the **Application settings** as follows:
+    * `ElasticServerlessForwarderS3ConfigFile` set the value of the S3 URL in the format `s3://bucket-name/config-file-name` to the configuration file for your deployment of Elastic Serverless Forwarder - this will populate the `S3_CONFIG_FILE` environment variable of the deployed Lambda.
```
Ah, yes - you're right, that explicit step is needed here and in general the Deploy/Configure split is a little forced/awkward. 

I think maybe the original 2 step process in deploy- needs to be 3 steps:

* Install {aws} integration assets in {kib}
* Create and upload `config.yaml`
* Deploy `elastic-serverless-forwarder` from Serverless Application Repository

ollyhowell commented:
@aspacca do you know if the instructions in https://www.elastic.co/blog/elastic-and-aws-serverless-application-repository-speed-time-to-actionable-insights-with-frictionless-log-ingestion-from-amazon-s3 are needed / still relevant?

i.e. steps 2 and 3 in **Before you begin**

ravikesarwani commented:
@ollyhowell Is the preview link (https://observability-docs_2214.docs-preview.app.elstc.co/guide/en/observability/master/aws-elastic-serverless-forwarder.html) up-to-date with all the changes so far? If not, can we please get it updated? Thanks!

ravikesarwani commented on docs/en/observability/deploy-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,296 @@
+image::images/aws-serverless-forwarder-install-assets.png[Find and install AWS integration assets in {kib}]
+
+Adding integrations from {kib} provides appropriate pre-built dashboards, ingest node configurations, and other assets that help you get the most out of the data you ingest. The integrations use https://www.elastic.co/guide/en/elasticsearch/reference/current/data-streams.html[data streams] with specific https://www.elastic.co/blog/an-introduction-to-the-elastic-data-stream-naming-scheme[naming co ...
+
```
Let's add a Note here. Integration assets are our preferred way but we support writing to any index or data stream.
Note: While Elastic recommends using the integration assets, the Lambda function also supports writing directly to an index, alias, or custom data stream. This enables existing Elasticsearch users to re-use index templates, ingest pipelines, or dashboards that are already created and connected to business processes. If you already have an index or data stream set up where you want to send the data, then you can skip the above step.

ollyhowell replied:
Added in next commit...

ollyhowell commented:
> @ollyhowell Is the preview link (https://observability-docs_2214.docs-preview.app.elstc.co/guide/en/observability/master/aws-elastic-serverless-forwarder.html) up-to-date with all the changes so far? If not, can we please get it updated? Thanks!

Yes, this has the latest changes.

aspacca commented:
> i.e. steps 2 and 3 in **Before you begin**

yes, they are still needed/relevant, but specifically to the blog topic of vpcflow logs ingestion

ollyhowell commented:
> > Yes, this has the latest changes.
> 
> nit: https://observability-docs_2214.docs-preview.app.elstc.co/guide/en/observability/master/images/multiline-regexp-test-repl-main.png https://observability-docs_2214.docs-preview.app.elstc.co/guide/en/observability/master/images/multiline-regexp-test-repl-run.png
> 
> are still the old screenshots :)

Yeah, I've not added your new images yet! Unless you've pushed them and we just need to prompt the build again?

aspacca commented:
> Unless you've pushed them and we just need to prompt the build again?

No, I didn't :)

ravikesarwani commented:
Hi @ollyhowell, What's the latest on getting this published?

My take on "Cut back content in https://github.com/elastic/elastic-serverless-forwarder/blob/main/docs/README-AWS.md? Work out overlap with what needs to be integrated within AWS docs (currently whole README!)" is that we keep a very basic introductory content available in the github.  Give them an overview of what this is, list of supported inputs, and then point them to the observability docs for more details.  This way we keep the main content in one place and focus on continuous improvements and updates as we add new features and functionality.

aspacca commented:
@ollyhowell anything missing here?

btw, v1.5.0 of the lambda was published a couple of weeks ago with a new config setting for elasticsearch output
```
args.ssl_assert_fingerprint: SSL fingerprint for self-signed SSL certificate on HTTPS transport
```

ollyhowell commented:
> btw, v1.5.0 of the lambda was published a couple of weeks ago with a new config setting for elasticsearch output
> 
> ```
> args.ssl_assert_fingerprint: SSL fingerprint for self-signed SSL certificate on HTTPS transport
> ```

Should this be added to the Fields section on /deploy?
i.e.
> inputs.[].outputs.[].args: Custom init arguments for the specified forwarding target output.
> For elasticsearch the following arguments are supported:
> args.elasticsearch_url etc.

ollyhowell commented:
@elasticmachine run elasticsearch-ci/docs rebuild

aspacca commented:
> Should this be added to the Fields section on /deploy?
> i.e.

exactly!

ollyhowell commented:
After finally getting this working and pushing some content changes (build pending...), quick update/recap:

- based on feedback and own experience, have merged the configure and transform topics together into a Configuration options topic (and corrected some internal filenames during process); although longer pages it makes more sense as the deploy-configure-transform division doesn't quite make sense for docs and the order we expect users to do things. This outline/organization will likely change in near future anyway when we tackle the other cloud monitoring / serverless options and update the Send Data section so I don't think we should stress too much about it right now (main goal is to stop sending users to the GH readme when we have a better version of docs!)
- added a new Pre-requisites section to the Deploy topic (worth reviewing and expanding!)
- have added the new configuration arg for SSL
- have expanded the Troubleshooting section, inc. Debug environment variable; also a note on SQS timeout increase
- added a note with Deploy using Cloudformation section to advise using same Application name to update existing function
- created a related issue for a dedicated tutorial for this topic (this content has some instructional steps but is not quite an end-to-end tutorial): https://github.com/elastic/observability-docs/issues/2370
- updated Multiline screenshots

Some outstanding questions/notes (for @aspacca in particular):
- do we have to update the code comments (and also field documentation) re: username/password vs. api_key precedence?
- left an internal comment looking for more details on pre-reqs for other input types e.g. Cloudwatch, Firehose
- in general are we happy to refer to this as *the forwarder* (after initial references with full Elastic Serverless Forwarder) rather than e.g. *The Lambda function* (used throughout top level topic)
- do we have a definitive step we want to recommend to the user to confirm logs are received? I just used the Logs view but perhaps we can be precise here i.e. use this view and filter by X

aspacca commented on docs/en/observability/aws-deploy-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,317 @@
+
+`inputs.[].outputs`:
+
+A list of outputs (i.e. forwarding targets) for the Elastic Serverless Forwarder Lambda function. Only one output can be defined per type.
```
```suggestion
A list of outputs (i.e. forwarding targets) for the Elastic Serverless Forwarder Lambda function. You can have multiple outputs for an input, but only one output can be defined per type.
```

aspacca commented on docs/en/observability/aws-deploy-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,317 @@
+
+  * `args.elasticsearch_url`: URL of elasticsearch endpoint in the format `http(s)://domain.tld:port`. Mandatory when `args.cloud_id` is not provided. Will take precedence over `args.cloud_id` if both are defined.
+  * `args.cloud_id`: Cloud ID of elasticsearch endpoint. Mandatory when `args.elasticsearch_url` is not provided. Will be ignored if `args.elasticsearch_url` is defined as well.
+  * `args.username`: Username of the elasticsearch instance to connect to. Mandatory when `args.api_key` is not provided. Will be ignored if `args.api_key` is defined as well.
```
```suggestion
  * `args.username`: Username of the elasticsearch instance to connect to. Mandatory when `args.api_key` is not provided. Will take precedence over `args.api_key` if both are defined.
```

aspacca commented on docs/en/observability/aws-deploy-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,317 @@
+  * `args.elasticsearch_url`: URL of elasticsearch endpoint in the format `http(s)://domain.tld:port`. Mandatory when `args.cloud_id` is not provided. Will take precedence over `args.cloud_id` if both are defined.
+  * `args.cloud_id`: Cloud ID of elasticsearch endpoint. Mandatory when `args.elasticsearch_url` is not provided. Will be ignored if `args.elasticsearch_url` is defined as well.
+  * `args.username`: Username of the elasticsearch instance to connect to. Mandatory when `args.api_key` is not provided. Will be ignored if `args.api_key` is defined as well.
+  * `args.password` Password of the elasticsearch instance to connect to. Mandatory when `args.api_key` is not provided. Will be ignored if `args.api_key` is defined as well.
```
```suggestion
  * `args.password` Password of the elasticsearch instance to connect to. Mandatory when `args.api_key` is not provided. Will take precedence over `args.api_key` if both are defined.
```

aspacca commented on docs/en/observability/aws-deploy-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,317 @@
+  * `args.cloud_id`: Cloud ID of elasticsearch endpoint. Mandatory when `args.elasticsearch_url` is not provided. Will be ignored if `args.elasticsearch_url` is defined as well.
+  * `args.username`: Username of the elasticsearch instance to connect to. Mandatory when `args.api_key` is not provided. Will be ignored if `args.api_key` is defined as well.
+  * `args.password` Password of the elasticsearch instance to connect to. Mandatory when `args.api_key` is not provided. Will be ignored if `args.api_key` is defined as well.
+  * `args.api_key`:  API key of elasticsearch endpoint in the format **base64encode(api_key_id:api_key_secret)**. Mandatory when `args.username`  and `args.password` are not provided. Will take precedence over `args.username`/`args.password` if both are defined.
```
```suggestion
  * `args.api_key`:  API key of elasticsearch endpoint in the format **base64encode(api_key_id:api_key_secret)**. Mandatory when `args.username`  and `args.password` are not provided. Will be ignored if `args.username`/`args.password` are defined as well.
```

aspacca commented on docs/en/observability/aws-deploy-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,317 @@
+  * `args.password` Password of the elasticsearch instance to connect to. Mandatory when `args.api_key` is not provided. Will be ignored if `args.api_key` is defined as well.
+  * `args.api_key`:  API key of elasticsearch endpoint in the format **base64encode(api_key_id:api_key_secret)**. Mandatory when `args.username`  and `args.password` are not provided. Will take precedence over `args.username`/`args.password` if both are defined.
+  * `args.es_datastream_name`: Name of data stream or index where logs should be forwarded to. Lambda supports automatic routing of various {aws} service logs to the corresponding data streams for further processing and storage in the {es} cluster. It supports automatic routing of `aws.cloudtrail`, `aws.cloudwatch_logs`, `aws.elb_logs`, `aws.firewall_logs`, `aws.vpcflow`, and `aws.waf` logs. For  ...
+  * `args.batch_max_actions`: Maximum number of actions to send in a single bulk request. Default value: 500.
```
```suggestion
  * `args.batch_max_actions`: (Optional) Maximum number of actions to send in a single bulk request. Default value: 500.
```

aspacca commented on docs/en/observability/aws-deploy-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,317 @@
+  * `args.api_key`:  API key of elasticsearch endpoint in the format **base64encode(api_key_id:api_key_secret)**. Mandatory when `args.username`  and `args.password` are not provided. Will take precedence over `args.username`/`args.password` if both are defined.
+  * `args.es_datastream_name`: Name of data stream or index where logs should be forwarded to. Lambda supports automatic routing of various {aws} service logs to the corresponding data streams for further processing and storage in the {es} cluster. It supports automatic routing of `aws.cloudtrail`, `aws.cloudwatch_logs`, `aws.elb_logs`, `aws.firewall_logs`, `aws.vpcflow`, and `aws.waf` logs. For  ...
+  * `args.batch_max_actions`: Maximum number of actions to send in a single bulk request. Default value: 500.
+  * `args.batch_max_bytes`: Maximum size in bytes to send in a single bulk request. Default value: 10485760 (10MB).
```
```suggestion
  * `args.batch_max_bytes`: (Optional) Maximum size in bytes to send in a single bulk request. Default value: 10485760 (10MB).
```

aspacca commented on docs/en/observability/aws-deploy-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,317 @@
+  * `args.es_datastream_name`: Name of data stream or index where logs should be forwarded to. Lambda supports automatic routing of various {aws} service logs to the corresponding data streams for further processing and storage in the {es} cluster. It supports automatic routing of `aws.cloudtrail`, `aws.cloudwatch_logs`, `aws.elb_logs`, `aws.firewall_logs`, `aws.vpcflow`, and `aws.waf` logs. For  ...
+  * `args.batch_max_actions`: Maximum number of actions to send in a single bulk request. Default value: 500.
+  * `args.batch_max_bytes`: Maximum size in bytes to send in a single bulk request. Default value: 10485760 (10MB).
+  * `args.ssl_assert_fingerprint`: SSL fingerprint for self-signed SSL certificate on HTTPS transport.
```
```suggestion
  * `args.ssl_assert_fingerprint`: (Optional) SSL fingerprint for self-signed SSL certificate on HTTPS transport.
```

aspacca commented on docs/en/observability/aws-deploy-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,317 @@
+  * `args.batch_max_bytes`: Maximum size in bytes to send in a single bulk request. Default value: 10485760 (10MB).
+  * `args.ssl_assert_fingerprint`: SSL fingerprint for self-signed SSL certificate on HTTPS transport.
+
+// is there some logic to this order? i.e. just placed ssl at end - should we re-order?
```
no logic but for readability one

aspacca commented on docs/en/observability/aws-deploy-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,317 @@
+=== Deploy using {aws} Console
+
+. Log in to {aws} console and navigate to the Lambda service.
+. Click **Create function**.
```
this changed a little in the current AWS console :(

ollyhowell replied:
I ran thru it again and updated the text and screenshot

aspacca commented on docs/en/observability/aws-deploy-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,317 @@
+[role="screenshot"]
+image::images/aws-serverless-forwarder-create-function.png[Create Elastic Serverless Forwarder Lambda function within SAR]
++
+. Complete the **Application settings** as follows:
```
let's find a way to clarify that most of the ARNs to provide are all the ones referenced in the config file.
Only `ElasticServerlessForwarderS3Buckets` are the ARNs of the s3 bucket of the set up sqs s3 notification that are not referenced in the config file, and `ElasticServerlessForwarderKMSKeys` are the ARNs of the KMS keys used to encrypt the SSM secrets that are not referenced in the config file.

ollyhowell replied:
I've updated the initial sentence to
`Complete the **Application settings** as follows, making sure you reference the ARNs specified in your `config.yaml`, and leaving any settings for unused inputs blank:` 
and simplified the text within the individual settings.

aspacca commented on docs/en/observability/aws-deploy-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,317 @@
+    * `ElasticServerlessForwarderS3ConfigFile`: Set the value of the S3 URL in the format `s3://bucket-name/config-file-name` to the configuration file for your deployment of Elastic Serverless Forwarder. This will populate the `S3_CONFIG_FILE` environment variable of the deployed Lambda.
+    * `ElasticServerlessForwarderSSMSecrets`: Add a comma delimited list of {aws} SSM Secrets ARNs referenced in the `config.yaml` file (if any).
+    * `ElasticServerlessForwarderKMSKeys`: Add a comma delimited list of {aws} KMS Keys ARNs to be used for decrypting {aws} SSM Secrets referenced in the `config.yaml` file (if any).
+    * `ElasticServerlessForwarderSQSEvents`: Add a comma delimited list of Direct SQS queues ARNs to set as event triggers for the Lambda (if any).
```
```suggestion
    * `ElasticServerlessForwarderSQSEvents`: Add a comma delimited list of Direct SQS queues ARNs to set as event triggers for the Lambda referenced in the `config.yaml` file (if any).
```

aspacca commented on docs/en/observability/aws-deploy-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,317 @@
+    * `ElasticServerlessForwarderSSMSecrets`: Add a comma delimited list of {aws} SSM Secrets ARNs referenced in the `config.yaml` file (if any).
+    * `ElasticServerlessForwarderKMSKeys`: Add a comma delimited list of {aws} KMS Keys ARNs to be used for decrypting {aws} SSM Secrets referenced in the `config.yaml` file (if any).
+    * `ElasticServerlessForwarderSQSEvents`: Add a comma delimited list of Direct SQS queues ARNs to set as event triggers for the Lambda (if any).
+    * `ElasticServerlessForwarderS3SQSEvents`: Add a comma delimited list of S3 SQS Event Notifications ARNs to set as event triggers for the Lambda (if any).
```
```suggestion
    * `ElasticServerlessForwarderS3SQSEvents`: Add a comma delimited list of S3 SQS Event Notifications ARNs to set as event triggers for the Lambda referenced in the `config.yaml` file (if any).
```

aspacca commented on docs/en/observability/aws-deploy-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,317 @@
+    * `ElasticServerlessForwarderKMSKeys`: Add a comma delimited list of {aws} KMS Keys ARNs to be used for decrypting {aws} SSM Secrets referenced in the `config.yaml` file (if any).
+    * `ElasticServerlessForwarderSQSEvents`: Add a comma delimited list of Direct SQS queues ARNs to set as event triggers for the Lambda (if any).
+    * `ElasticServerlessForwarderS3SQSEvents`: Add a comma delimited list of S3 SQS Event Notifications ARNs to set as event triggers for the Lambda (if any).
+    * `ElasticServerlessForwarderKinesisEvents`: Add a comma delimited list of Kinesis Data Stream ARNs to set as event triggers for the Lambda (if any).
```
```suggestion
    * `ElasticServerlessForwarderKinesisEvents`: Add a comma delimited list of Kinesis Data Stream ARNs to set as event triggers for the Lambda referenced in the `config.yaml` file (if any).
```

aspacca commented on docs/en/observability/aws-deploy-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,317 @@
+    * `ElasticServerlessForwarderSQSEvents`: Add a comma delimited list of Direct SQS queues ARNs to set as event triggers for the Lambda (if any).
+    * `ElasticServerlessForwarderS3SQSEvents`: Add a comma delimited list of S3 SQS Event Notifications ARNs to set as event triggers for the Lambda (if any).
+    * `ElasticServerlessForwarderKinesisEvents`: Add a comma delimited list of Kinesis Data Stream ARNs to set as event triggers for the Lambda (if any).
+    * `ElasticServerlessForwarderCloudWatchLogsEvents`: Add a comma delimited list of Cloudwatch Logs log group ARNs to set subscription filters on the Lambda (if any).
```
```suggestion
    * `ElasticServerlessForwarderCloudWatchLogsEvents`: Add a comma delimited list of Cloudwatch Logs log group ARNs to set subscription filters on the Lambda referenced in the `config.yaml` file (if any).
```

aspacca commented on docs/en/observability/aws-deploy-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,317 @@
+            ElasticServerlessForwarderS3ConfigFile: ""          ## FILL WITH THE VALUE OF THE S3 URL IN THE FORMAT "s3://bucket-name/config-file-name" POINTING TO THE CONFIGURATION FILE FOR YOUR DEPLOYMENT OF THE ELASTIC SERVERLESS FORWARDER
+            ElasticServerlessForwarderSSMSecrets: ""            ## FILL WITH A COMMA DELIMITED LIST OF AWS SSM SECRETS ARNS REFERENCED IN THE CONFIG YAML FILE (IF ANY).
+            ElasticServerlessForwarderKMSKeys: ""               ## FILL WITH A COMMA DELIMITED LIST OF AWS KMS KEYS ARNS TO BE USED FOR DECRYPTING AWS SSM SECRETS REFERENCED IN THE CONFIG YAML FILE (IF ANY).
+            ElasticServerlessForwarderSQSEvents: ""             ## FILL WITH A COMMA DELIMITED LIST OF DIRECT SQS QUEUES ARNS TO SET AS EVENT TRIGGERS FOR THE LAMBDA (IF ANY).
```
```suggestion
            ElasticServerlessForwarderSQSEvents: ""             ## FILL WITH A COMMA DELIMITED LIST OF DIRECT SQS QUEUES ARNS TO SET AS EVENT TRIGGERS FOR THE LAMBDA REFERENCED IN THE CONFIG YAML FILE (IF ANY).
```

aspacca commented on docs/en/observability/aws-deploy-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,317 @@
+            ElasticServerlessForwarderSSMSecrets: ""            ## FILL WITH A COMMA DELIMITED LIST OF AWS SSM SECRETS ARNS REFERENCED IN THE CONFIG YAML FILE (IF ANY).
+            ElasticServerlessForwarderKMSKeys: ""               ## FILL WITH A COMMA DELIMITED LIST OF AWS KMS KEYS ARNS TO BE USED FOR DECRYPTING AWS SSM SECRETS REFERENCED IN THE CONFIG YAML FILE (IF ANY).
+            ElasticServerlessForwarderSQSEvents: ""             ## FILL WITH A COMMA DELIMITED LIST OF DIRECT SQS QUEUES ARNS TO SET AS EVENT TRIGGERS FOR THE LAMBDA (IF ANY).
+            ElasticServerlessForwarderS3SQSEvents: ""           ## FILL WITH A COMMA DELIMITED LIST OF S3 SQS EVENT NOTIFICATIONS ARNS TO SET AS EVENT TRIGGERS FOR THE LAMBDA (IF ANY).
```
```suggestion
            ElasticServerlessForwarderS3SQSEvents: ""           ## FILL WITH A COMMA DELIMITED LIST OF S3 SQS EVENT NOTIFICATIONS ARNS TO SET AS EVENT TRIGGERS FOR THE LAMBDA REFERENCED IN THE CONFIG YAML FILE (IF ANY).
```

aspacca commented on docs/en/observability/aws-deploy-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,317 @@
+            ElasticServerlessForwarderKMSKeys: ""               ## FILL WITH A COMMA DELIMITED LIST OF AWS KMS KEYS ARNS TO BE USED FOR DECRYPTING AWS SSM SECRETS REFERENCED IN THE CONFIG YAML FILE (IF ANY).
+            ElasticServerlessForwarderSQSEvents: ""             ## FILL WITH A COMMA DELIMITED LIST OF DIRECT SQS QUEUES ARNS TO SET AS EVENT TRIGGERS FOR THE LAMBDA (IF ANY).
+            ElasticServerlessForwarderS3SQSEvents: ""           ## FILL WITH A COMMA DELIMITED LIST OF S3 SQS EVENT NOTIFICATIONS ARNS TO SET AS EVENT TRIGGERS FOR THE LAMBDA (IF ANY).
+            ElasticServerlessForwarderKinesisEvents: ""         ## FILL WITH A COMMA DELIMITED LIST OF KINESIS DATA STREAM ARNS TO SET AS EVENT TRIGGERS FOR THE LAMBDA (IF ANY).
```
```suggestion
            ElasticServerlessForwarderKinesisEvents: ""         ## FILL WITH A COMMA DELIMITED LIST OF KINESIS DATA STREAM ARNS TO SET AS EVENT TRIGGERS FOR THE LAMBDA REFERENCED IN THE CONFIG YAML FILE (IF ANY).
```

aspacca commented on docs/en/observability/aws-deploy-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,317 @@
+            ElasticServerlessForwarderSQSEvents: ""             ## FILL WITH A COMMA DELIMITED LIST OF DIRECT SQS QUEUES ARNS TO SET AS EVENT TRIGGERS FOR THE LAMBDA (IF ANY).
+            ElasticServerlessForwarderS3SQSEvents: ""           ## FILL WITH A COMMA DELIMITED LIST OF S3 SQS EVENT NOTIFICATIONS ARNS TO SET AS EVENT TRIGGERS FOR THE LAMBDA (IF ANY).
+            ElasticServerlessForwarderKinesisEvents: ""         ## FILL WITH A COMMA DELIMITED LIST OF KINESIS DATA STREAM ARNS TO SET AS EVENT TRIGGERS FOR THE LAMBDA (IF ANY).
+            ElasticServerlessForwarderCloudWatchLogsEvents: ""  ## FILL WITH A COMMA DELIMITED LIST OF CLOUDWATCH LOGS LOG GROUPS ARNS TO SET SUBSCRIPTION FILTERS ON THE LAMBDA FOR (IF ANY).
```
```suggestion
            ElasticServerlessForwarderCloudWatchLogsEvents: ""  ## FILL WITH A COMMA DELIMITED LIST OF CLOUDWATCH LOGS LOG GROUPS ARNS TO SET SUBSCRIPTION FILTERS ON THE LAMBDA FOR REFERENCED IN THE CONFIG YAML FILE (IF ANY).
```

aspacca commented on docs/en/observability/aws-deploy-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,317 @@
+      ElasticServerlessForwarderS3ConfigFile         = ""  ## FILL WITH THE VALUE OF THE S3 URL IN THE FORMAT "s3://bucket-name/config-file-name" POINTING TO THE CONFIGURATION FILE FOR YOUR DEPLOYMENT OF THE ELASTIC SERVERLESS FORWARDER
+      ElasticServerlessForwarderSSMSecrets           = ""  ## FILL WITH A COMMA DELIMITED LIST OF AWS SSM SECRETS ARNS REFERENCED IN THE CONFIG YAML FILE (IF ANY).
+      ElasticServerlessForwarderKMSKeys              = ""  ## FILL WITH A COMMA DELIMITED LIST OF AWS KMS KEYS ARNS TO BE USED FOR DECRYPTING AWS SSM SECRETS REFERENCED IN THE CONFIG YAML FILE (IF ANY).
+      ElasticServerlessForwarderSQSEvents            = ""  ## FILL WITH A COMMA DELIMITED LIST OF DIRECT SQS QUEUES ARNS TO SET AS EVENT TRIGGERS FOR THE LAMBDA (IF ANY).
```
```suggestion
      ElasticServerlessForwarderSQSEvents            = ""  ## FILL WITH A COMMA DELIMITED LIST OF DIRECT SQS QUEUES ARNS TO SET AS EVENT TRIGGERS FOR THE LAMBDA REFERENCED IN THE CONFIG YAML FILE (IF ANY).
```

aspacca commented on docs/en/observability/aws-deploy-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,317 @@
+      ElasticServerlessForwarderSSMSecrets           = ""  ## FILL WITH A COMMA DELIMITED LIST OF AWS SSM SECRETS ARNS REFERENCED IN THE CONFIG YAML FILE (IF ANY).
+      ElasticServerlessForwarderKMSKeys              = ""  ## FILL WITH A COMMA DELIMITED LIST OF AWS KMS KEYS ARNS TO BE USED FOR DECRYPTING AWS SSM SECRETS REFERENCED IN THE CONFIG YAML FILE (IF ANY).
+      ElasticServerlessForwarderSQSEvents            = ""  ## FILL WITH A COMMA DELIMITED LIST OF DIRECT SQS QUEUES ARNS TO SET AS EVENT TRIGGERS FOR THE LAMBDA (IF ANY).
+      ElasticServerlessForwarderS3SQSEvents          = ""  ## FILL WITH A COMMA DELIMITED LIST OF S3 SQS EVENT NOTIFICATIONS ARNS TO SET AS EVENT TRIGGERS FOR THE LAMBDA (IF ANY).
```
```suggestion
      ElasticServerlessForwarderS3SQSEvents          = ""  ## FILL WITH A COMMA DELIMITED LIST OF S3 SQS EVENT NOTIFICATIONS ARNS TO SET AS EVENT TRIGGERS FOR THE LAMBDA REFERENCED IN THE CONFIG YAML FILE (IF ANY).
```

aspacca commented on docs/en/observability/aws-deploy-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,317 @@
+      ElasticServerlessForwarderKMSKeys              = ""  ## FILL WITH A COMMA DELIMITED LIST OF AWS KMS KEYS ARNS TO BE USED FOR DECRYPTING AWS SSM SECRETS REFERENCED IN THE CONFIG YAML FILE (IF ANY).
+      ElasticServerlessForwarderSQSEvents            = ""  ## FILL WITH A COMMA DELIMITED LIST OF DIRECT SQS QUEUES ARNS TO SET AS EVENT TRIGGERS FOR THE LAMBDA (IF ANY).
+      ElasticServerlessForwarderS3SQSEvents          = ""  ## FILL WITH A COMMA DELIMITED LIST OF S3 SQS EVENT NOTIFICATIONS ARNS TO SET AS EVENT TRIGGERS FOR THE LAMBDA (IF ANY).
+      ElasticServerlessForwarderKinesisEvents        = ""  ## FILL WITH A COMMA DELIMITED LIST OF KINESIS DATA STREAM ARNS TO SET AS EVENT TRIGGERS FOR THE LAMBDA (IF ANY).
```
```suggestion
      ElasticServerlessForwarderKinesisEvents        = ""  ## FILL WITH A COMMA DELIMITED LIST OF KINESIS DATA STREAM ARNS TO SET AS EVENT TRIGGERS FOR THE LAMBDA REFERENCED IN THE CONFIG YAML FILE (IF ANY).
```

aspacca commented on docs/en/observability/aws-deploy-elastic-serverless-forwarder.asciidoc:
```diff
@@ -0,0 +1,317 @@
+      ElasticServerlessForwarderSQSEvents            = ""  ## FILL WITH A COMMA DELIMITED LIST OF DIRECT SQS QUEUES ARNS TO SET AS EVENT TRIGGERS FOR THE LAMBDA (IF ANY).
+      ElasticServerlessForwarderS3SQSEvents          = ""  ## FILL WITH A COMMA DELIMITED LIST OF S3 SQS EVENT NOTIFICATIONS ARNS TO SET AS EVENT TRIGGERS FOR THE LAMBDA (IF ANY).
+      ElasticServerlessForwarderKinesisEvents        = ""  ## FILL WITH A COMMA DELIMITED LIST OF KINESIS DATA STREAM ARNS TO SET AS EVENT TRIGGERS FOR THE LAMBDA (IF ANY).
+      ElasticServerlessForwarderCloudWatchLogsEvents = ""  ## FILL WITH A COMMA DELIMITED LIST OF CLOUDWATCH LOGS LOG GROUPS ARNS TO SET SUBSCRIPTION FILTERS ON THE LAMBDA FOR (IF ANY).
```
```suggestion
      ElasticServerlessForwarderCloudWatchLogsEvents = ""  ## FILL WITH A COMMA DELIMITED LIST OF CLOUDWATCH LOGS LOG GROUPS ARNS TO SET SUBSCRIPTION FILTERS ON THE LAMBDA FOR REFERENCED IN THE CONFIG YAML FILE (IF ANY).
```

aspacca reviewed (commented):
suggesting some minor changes

ollyhowell commented:
@elasticmachine run elasticsearch-ci/docs rebuild

ravikesarwani commented:
> in general are we happy to refer to this as the forwarder (after initial references with full Elastic Serverless Forwarder) rather than e.g. The Lambda function (used throughout top level topic) -- what do you think @ravikesarwani ?

@ollyhowell Yes, I agree. Let's call it "forwarder" after initial references with a full "Elastic Serverless Forwarder" name (and the first reference on each new page). I think that will make it consistent and easier for readers. In the first line, we introduce it as a Lambda function so I think we are covered there.

This is looking pretty good! Nick work. Thank you Olly, all the doc team members, and Andrea for your contributions and for making it simpler and easier for our users to follow along.

ollyhowell commented:
Thanks @ravikesarwani @aspacca 
Will push a minor change soon with some tidying of references to The Lambda function, the forwarder etc. but can I get an approve from one of you if you're happy with the PR in general? Or feel free to comment further or wait for next commit if you prefer, just want to get this ready to merge this week when we're happy.

aspacca reviewed (approved):
minor: comments on params for cloudformation and terraform templates still sometimes mention "referenced in your config yaml" and sometimes not.
let's go with the same format for every param

ollyhowell commented:
Thank you @aspacca 
Tidied the code comments references to config file (and the 'function' rather than the 'lambda') and pushed new version.

Which versions should this content be backported to?
Are you ok with me proposing the changes to the GitHub readme? i.e. cutting back the repo version pulled into SAR

aspacca commented:
> Which versions should this content be backported to?

this is version 1.5.0 version of the forwarder on aws (we will probably have different version of the forwarder according to the cloud provider, in the future)



> Are you ok with me proposing the changes to the GitHub readme? i.e. cutting back the repo version pulled into SAR

I would totally remove the readme on github, and just reference the official docs.
new docs will go directly through a PR on observability-docs repo, readme content on SAR will present a brief description with a link to the official docs.

while this will require to docs to be updated/published before realising a new version on SAR, it's better to have a single source of truth in my opinion

ollyhowell commented:
Sorry, I meant which versions of the Elastic docs should we backport to?

For the README the plan was to cut back to minimum and link thru to official docs - we can't remove completely as this content is pulled into SAR so I will propose a couple of paragraphs and bullets and link to our page (once live). I will tag you and @ravikesarwani on the PR for this anyway.

aspacca commented:
> Sorry, I meant which versions of the Elastic docs should we backport to?

not sure about this. forwarder versions/releases are out of band compared to stack versions/releases. the forwarder is backward compatibile since `7.16` on Elasticsearch, but this might differ for Logstash once we introduce the output.

aspacca commented:
> For the README the plan was to cut back to minimum and link thru to official docs - we can't remove completely as this content is pulled into SAR so I will propose a couple of paragraphs and bullets and link to our page (once live)

Beware that we must release a new forwarder version to update the README pulled into on SAR.

I miss how relations between Elastic docs version and the forwarder version will work: we could need to update forwarder docs without any coupling to a stack version, does this mean we cannot publish it until a new stack version is released?

ollyhowell commented:
> not sure about this. forwarder versions/releases are out of band compared to stack versions/releases. the forwarder is backward compatibile since `7.16` on Elasticsearch, but this might differ for Logstash once we introduce the output.

I would suggest we backpot to 7.16 then and then further changes for Logstash will be backported as appropriate.

ollyhowell commented:
Need to reopen this and backport to 8.6 - think it must have been merged and backported just after 8.6 branch was created and I missed adding the backport-8.6 label for mergify...

dedemorton commented:
@ollyhowell The automation is so awesome that you can just add the backport label even if the PR has been merged! You should check the PR and make sure it's good to go.
